lib/configurators/aws/aws.go

```diff
@@ -499,19 +504,64 @@ func policiesTarget(flags configurators.BootstrapFlags, accountID string, partit
 	if flags.AttachToRole != "" {
 		roleArn := flags.AttachToRole
 		if !arn.IsARN(flags.AttachToRole) {
-			roleArn = fmt.Sprintf("arn:%s:iam::%s:role/%s", partitionID, accountID, flags.AttachToRole)
+			roleArn = buildIAMARN(partitionID, accountID, "role", flags.AttachToRole)
 		}
 
 		return awslib.IdentityFromArn(roleArn)
 	}
 
 	if identity == nil {
-		return awslib.IdentityFromArn(fmt.Sprintf("arn:%s:iam::%s:user/%s", partitionID, accountID, defaultAttachUser))
+		return awslib.IdentityFromArn(buildIAMARN(partitionID, accountID, "user", defaultAttachUser))
+	}
+
+	if identity.GetType() == "assumed-role" {
+		roleIdentity, err := getRoleARNForAssumedRole(iamClient, identity)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return roleIdentity, nil
 	}
 
 	return identity, nil
 }
 
+// buildIAMARN constructs an AWS IAM ARN string from the given partition,
+// account, resource type, and resource.
+// If the resource starts with the "/" prefix, this function takes care not to
+// add an additional "/" prefix to the constructed ARN.
+// This handles resource names that include a path correctly. Example:
+// resource input: "/some/path/to/rolename"
+// arn output: "arn:aws:iam::123456789012:role/some/path/to/rolename"
+func buildIAMARN(partitionID, accountID, resourceType, resource string) string {
+	if strings.HasPrefix(resource, "/") {
+		return fmt.Sprintf("arn:%s:iam::%s:%s%s", partitionID, accountID, resourceType, resource)
+	} else {
+		return fmt.Sprintf("arn:%s:iam::%s:%s/%s", partitionID, accountID, resourceType, resource)
+	}
+}
+
+// failedToResolveAssumeRoleARN is an error message returned when an
+// assumed-role identity cannot be resolved to the role ARN that it assumes,
+// which is necessary to attach policies to the identity.
+// Rather than returning errors about why it failed, this message suggests a
+// simple fix for the user to specify a role or user to attach policies to.
+const failedToResolveAssumeRoleARN = "Running with assumed-role credentials. Policies cannot be attached to an assumed-role. Provide the name or ARN of the IAM user or role to attach policies to."
```

Comment:
`AttachRolePolicyInput` takes a "friendly" role name instead of the full ARN so technically policies can be attached if the target is an assumed role as only `identity.GetName()` is used.

Reply:
yes we only need the name to attach the policy. However, when we make IAM-edit statements (`iam:GetRolePolicy` `iam:PutRolePolicy` `iam:DeleteRolePolicy`), we need to use the full ARN of the role: https://github.com/gravitational/teleport/blob/3eafd129a94d83092762ca82fd7add29a1ee4b6e/lib/configurators/aws/aws.go#L859-L865

Reply:
👍 I see. Does that mean we have never caught this before for RDS/Redshift?

I'll approve as it is. Ideally, we should only call GetRole when necessary, or maybe convert assumed role identity into two role arns as resources here (like one can have a wildcard like role/*/<rolename> in case its a path)

Is iam client nil in manual mode?

Reply:
> Does that mean we have never caught this before for RDS/Redshift?

I guess so, although I think it's an uncommon scenario since typically the user is not going to do `teleport db configure bootstrap` without either `--attach-to-role` or `--attach-to-user`, as the configurator would need permissions to bootstrap its own permissions.

> Ideally, we should only call GetRole when necessary, or maybe convert assumed role identity into two role arns as resources in buildIAMEditStatements (like one resource can have a wildcard like role/*/ in case its a path)

I thought about that too, but then I realized this actually solves a problem for the aws cross-account PR, since an assumed role ARN won't be equal to `assume_role_arn`, which would mess up the partitioning of resources for the target identity. So I think we should just always do it when the target identity is an assumed role.

> Is iam client nil in manual mode? I think we should make sure we don't make api calls in manual mode.

it is, but so is identity, so it doesn't try to make the api call:

https://github.com/gravitational/teleport/blob/2e40609f65dba9761340573f71a61108dcca0d6e/lib/configurators/aws/aws.go#L513-L518

Reply:
> this actually solves a problem for the aws cross-account PR, since an assumed role ARN won't be equal to assume_role_arn,

Agreed the current approach makes this check easier. Alternatively, a custom check can be done when identity is a role, just compare the account id and GetName, without the full ARN